models/service_helpers.go

```diff
@@ -262,19 +262,40 @@ func RemoveService(q *reform.Querier, id string, mode RemoveMode) error {
 	if err != nil {
 		return err
 	}
-	// check/remove Agents
-	structs, err := q.FindAllFrom(AgentTable, "service_id", id)
+	// find agents and artifacts
+	agents, err := FindAgents(q, AgentFilters{ServiceID: id})
 	if err != nil {
 		return errors.Wrap(err, "failed to select Agent IDs")
 	}
-	if len(structs) != 0 {
+
+	artifacts, err := FindArtifacts(q, &ArtifactFilters{ServiceID: id})
+	if err != nil {
+		return errors.Wrap(err, "failed to select artifacts")
+	}
+
+	restoreItems, err := FindRestoreHistoryItems(q, &RestoreHistoryItemFilters{ServiceID: id})
+	if err != nil {
+		return errors.Wrap(err, "failed to select restore history items")
+	}
+
+	if len(agents) != 0 || len(artifacts) != 0 || len(restoreItems) != 0 {
 		switch mode {
 		case RemoveRestrict:
-			return status.Errorf(codes.FailedPrecondition, "Service with ID %q has agents.", id)
+			return status.Errorf(codes.FailedPrecondition,
+				"Service with ID %q has agents, artifacts or restore history items.", id)
 		case RemoveCascade:
-			for _, str := range structs {
-				agentID := str.(*Agent).AgentID
-				if _, err = RemoveAgent(q, agentID, RemoveCascade); err != nil {
+			for _, a := range agents {
+				if _, err := RemoveAgent(q, a.AgentID, RemoveCascade); err != nil {
+					return err
+				}
+			}
+			for _, a := range artifacts {
+				if err := RemoveArtifact(q, a.ID); err != nil {
+					return err
+				}
+			}
```

Comment:
Not sure that it's a good idea. Maybe we can just set empty `service_id` for such artifacts?

Reply:
I think I agree with Artem because of this use case.
1. User will backup service1
2. User deletes service1
3. User want restore backup from step(1) to new service.

Reply:
reworked, so now I'm returning empty ServiceID and ServiceName and keeping the artifacts for removed services